File: petshop-frontend/src/schemas.ts
import { z } from "zod";

export const PetSchema = z.object({
  id: z.number(),
  name: z.string(),
  species: z.string().nullable().optional(),
  owner_id: z.number(),
  age: z.number().nullable().optional(),
  breed: z.string().nullable().optional(),
  color: z.string().nullable().optional(),
  weight: z.number().nullable().optional(),
  description: z.string().nullable().optional(),
  gender: z.string().nullable().optional(),
  is_vaccinated: z.boolean().nullable().optional(),
  birthdate: z.string().nullable().optional(),
  date_added: z.string().nullable().optional(),
  photo_filename: z.string().nullable().optional(),
});

export const OwnerSchema = z.object({
  id: z.number(),
  name: z.string(),
  email: z.string().nullable().optional(),
  phone: z.string().nullable().optional(),
  address: z.string().nullable().optional(),
  city: z.string().nullable().optional(),
  state: z.string().nullable().optional(),
  zip_code: z.string().nullable().optional(),
  country: z.string().nullable().optional(),
  date_of_birth: z.string().nullable().optional(),
});

export const PetArraySchema = z.array(PetSchema);
export const OwnerArraySchema = z.array(OwnerSchema);

export type Pet = z.infer<typeof PetSchema>;
export type Owner = z.infer<typeof OwnerSchema>;

File: schemas.py
from pydantic import BaseModel, ConfigDict
from typing import List
from datetime import date, datetime


# Pet schemas
class PetCreate(BaseModel):
    name: str
    owner_id: int
    species: str | None = None
    age: int | None = None
    breed: str | None = None
    color: str | None = None
    weight: float | None = None
    description: str | None = None
    gender: str | None = None  # e.g. 'male', 'female', 'unknown'
    is_vaccinated: bool | None = None
    birthdate: date | None = None
    # photo_filename is not included here


class PetRead(BaseModel):
    id: int
    name: str
    owner_id: int
    species: str | None = None
    age: int | None = None
    breed: str | None = None
    color: str | None = None
    weight: float | None = None
    description: str | None = None
    gender: str | None = None
    is_vaccinated: bool | None = None
    birthdate: date | None = None
    date_added: datetime | None = None
    photo_filename: str | None = None

    model_config = ConfigDict(from_attributes=True)


# Owner schemas
class OwnerCreate(BaseModel):
    name: str
    email: str | None = None
    phone: str | None = None
    address: str | None = None
    city: str | None = None
    state: str | None = None
    zip_code: str | None = None
    country: str | None = None
    date_of_birth: str | None = None


class OwnerRead(BaseModel):
    id: int
    name: str
    email: str | None = None
    phone: str | None = None
    address: str | None = None
    city: str | None = None
    state: str | None = None
    zip_code: str | None = None
    country: str | None = None
    date_of_birth: str | None = None
    pets: List["PetRead"] = []

    model_config = ConfigDict(from_attributes=True)
